internal/cli/kraft/cloud/tunnel/tunnel.go

```diff
@@ -49,99 +88,317 @@ func NewCmd() *cobra.Command {
+				opts.exposedProxyPorts[i],
+				opts.Connections,
+			),
+		)
+	}
+
+	var allConnections string
+	for _, conn := range connections {
+		allConnections += conn + "|"
+	}
+	allConnections = "[" + strings.TrimSuffix(allConnections, "|") + "]"
+
+	return []string{
+		// HEARTBEAT_PORT
+		fmt.Sprintf("%d", opts.ProxyControlPort),
+		// AUTH_COOKIE_LEN:AUTH_TIMEOUT:AUTH_MAX_EVS:AUTH_COOKIE
+		fmt.Sprintf("%d:%d:%d:%s", len(authStr), 5, opts.Connections, authStr),
+		// MAX_EVS:GLOBAL_GPBUF:EVS_TIMEOUT
+		fmt.Sprintf("%d:%d:%d", opts.Connections, 8, 600),
+		// [HOOKSTR0:<HOOKSTR0_ARGS>|HOOKSTR1:<HOOKSTR1_ARGS>...]
+		allConnections,
+	}
+}
 
-	if len(ports) == 1 {
-		rport64, err := strconv.ParseUint(ports[0], 10, 16)
+// populatePrivateIPs fetches the private IPs of the instances and replaces the instance names/uuids with the Private IPs.
+func populatePrivateIPs(ctx context.Context, cli kcinstances.InstancesService, ips []string) ([]string, error) {
+	var instancesToGet []string
+	var indexesToGet []int
+	for i := range ips {
+		// If instance is not an IP (PrivateIP) or PrivateFQDN
+		// assume it is a name/UUID and fetch the IP
+		if net.ParseIP(ips[i]) == nil && !strings.HasSuffix(ips[i], ".internal") {
+			instancesToGet = append(instancesToGet, ips[i])
+			indexesToGet = append(indexesToGet, i)
+		}
+	}
+	if len(instancesToGet) > 0 {
+		instGetResp, err := cli.Get(ctx, instancesToGet...)
 		if err != nil {
-			return 0, 0, fmt.Errorf("%q is not a valid port number", ports[0])
+			return nil, fmt.Errorf("getting instances: %w", err)
+		}
+
+		for i, inst := range instGetResp.Data.Entries {
+			if inst.PrivateIP == "" {
+				return nil, fmt.Errorf("instance '%s' not found", instancesToGet[i])
+			}
+			ips[indexesToGet[i]] = inst.PrivateIP
 		}
-		return 0, uint16(rport64), nil
 	}
 
-	lport64, err := strconv.ParseUint(ports[0], 10, 16)
-	if err != nil {
-		return 0, 0, fmt.Errorf("%q is not a valid port number", ports[0])
+	return ips, nil
+}
+
+// runProxy runs a proxy instance with the given arguments.
+// Information related to the proxy instance is hardcoded, but the UUID is returned.
+func runProxy(ctx context.Context, cli kcinstances.InstancesService, proxyPorts []uint16, proxyControlPort uint16, args []string) (string, string, error) {
+	var parsedPorts []kcservices.CreateRequestService
+	for i := range proxyPorts {
+		parsedPorts = append(parsedPorts, kcservices.CreateRequestService{
+			Port:            int(proxyPorts[i]),
+			DestinationPort: ptr(int(proxyPorts[i])),
+			Handlers: []kcservices.Handler{
+				kcservices.HandlerTLS,
+			},
+		})
 	}
+	parsedPorts = append(parsedPorts, kcservices.CreateRequestService{
+		Port:            int(proxyControlPort),
+		DestinationPort: ptr(int(proxyControlPort)),
+		Handlers: []kcservices.Handler{
+			kcservices.HandlerTLS,
+		},
+	})
 
-	rport64, err := strconv.ParseUint(ports[1], 10, 16)
+	crinstResp, err := cli.Create(ctx, kcinstances.CreateRequest{
+		Image:    "kraftcloud/tunlr:latest",
```

Comment:
Let's define this image as a `const` and expose publicly as `TunnelServiceImage`.

Reply:
Additionally, this should be configurable as a CLI option, `---tunnel-service-image`.